rfcs/20200712-tfrt-kernel-fallback.md

```diff
@@ -0,0 +1,647 @@
+  ...
+};
+```
+
+When forwarding, we instantiate the kernel interfaces with TFRT’s lightweight
+OpKernel definitions, rather than TensorFlow’s
+[heavyweight OpKernel definitions](https://cs.opensource.google/android/platform/superproject/+/master:external/tensorflow/tensorflow/core/framework/op_kernel.h;l=612?q=opkernelcontext)
+for example.
+
+Example `AddN` kernel implementation using these new interfaces:
+
+```cpp
+class AddNOp : public OpKernelBase {
+ public:
+   explicit AddNOp(OpKernelConstructionInterface* construction) :
+       OpKernelBase(construction) {}
+
+   void Compute(OpKernelContextInterface* ctx) override {
+     if(!ctx->ValidateInputsAreSameShape(this)) return;
+     ...
+```
+
+Here, `OpKernelBase` implementation will be minimal:
+
+```cpp
+class OpKernelBase {
+ public:
+  explicit OpKernelBase(OpKernelConstructionInterface* context) {
+  }
+  virtual ~OpKernelBase() {}
+  virtual void Compute(OpKernelContextInterface* context) = 0;
+};
+```
+
+(For details how extending from `OpKernelBase` instead of `OpKernel` would work
+with current TensorFlow runtime see [Appendix 1](#appendix-1))
+
+Corresponding .cc file then registers the kernel using the correct kernel and
+context classes. For example, this is how we register `AddN` kernel with TFRT:
+
+```cpp
+REGISTER_KERNEL_FALLBACK_KERNEL( "AddN", AddNOp<CPUDevice, int32>);
+```
+
+## Calling kernel
+
+We add a new TFRT BEF kernel called `tfrt_fallback.kernel_fallback`. This kernel directly
+calls a TF kernel’s `Compute` method by creating `TFRTOpKernel*` data structures
+that forward to corresponding TFRT concepts. For example, the following code
+accesses an input in `llvm::ArrayRef<tfrt::RCReference<tfrt::AsyncValue>>` which
+we get from TFRT:
+
+```cpp
+const Tensor& TFRTOpKernelContext::input(int index) {
+  return inputs_[index]->get<Tensor>();
+}
+```
+
+Simplified definition of `tfrt_fallback.kernel_fallback`:
+
+```cpp
+// Instantiate a kernel. This would be a TensorFlow kernel converted to inherit
+// from `OpKernelBase` instead of `OpKernel`.
+std::unique_ptr<OpKernelBase> op  =  …;
+
+// Create TFRTOpKernelContext. The variable exec_ctx here is the tfrt::ExecutionContext passed to the kernel handler.
+TFRTOpKernelContext op_kernel_context(inputs, outputs.size(), op_meta, exec_ctx.host());
+
+// Directly invoke the TF kernel's Compute() method.
+op->Compute(&op_kernel_context);
+```
+
+## tfrt\_fallback.kernel\_fallback call structure
+
+We will be using the following conventions (essentially, these are based on
+Runtime Fallback work that will probably have RFC coming soon):
+
+*   Attributes are passed as key-value pairs, where both key and value are
+    represented as strings.
+*   Types have a specific string representation. We are trying to use names
+    consistent with BEF syntax as much as possible (for e.g. `f32` represents
+    `float`).
+*   Inputs and outputs have type `tensorflow::Tensor`. We will provide BEF
```

Comment:
which ops have this, other than tf.const (which we might want to just implement natively in TFRT)?